### PR TITLE
Allow transform with compensating offset change.

### LINKED ISSUE #93: Starting point shift with countering transform change should be ignored
Test case:
https://chromium-review.googlesource.com/c/chromium/src/+/2673965/6/third_party/blink/web_tests/external/wpt/layout-instability/transform-counter-layout-shift.html

This is a pattern used in many websites containing carousal UI, which moves elements smoothly using transform,
and when new contents is to be shifted in, resets the transform and add a countering layout offset, without any visual change.

### PR DESCRIPTION
**Pull request by skobes**
Fixes #93

***
<a href="https://pr-preview.s3.amazonaws.com/skobes/layout-instability/pull/94.html" title="Last updated on Feb 11, 2021, 7:50 PM UTC (8c923f4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/layout-instability/94/1cdfbe2...skobes:8c923f4.html" title="Last updated on Feb 11, 2021, 7:50 PM UTC (8c923f4)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -220,20 +220,23 @@ is defined as follows:
...
     <a>flow-relative</a> starting corner of the first <a>line box</a> generated
-    by |N|, calculated as if every <a>transformed element</a> had a
-    <a>transformation matrix</a> equal to the identity matrix.
+    by |N|.
 
-NOTE: The CSS transform is ignored for the calculation of the starting point,
-to ensure that a node is not made unstable by a transform change. However, the CSS
-transform is not ignored for the calculation of the visual representation and
-the associated exclusion of points outside of the viewport.
+The <dfn export>transform-indifferent starting point</dfn> of a {{Node}} |N| in a
```
Comment (wangxianzhu) [inline index.bs:230]: Just a side note: we may later need a "scroll-indifferent starting point" concept for ignoring scroll shift, and the counterpart of it for layout shift countering scroll.

File: index.bs [hunk 2]
```diff
@@ -257,17 +260,28 @@ The <dfn export>previous frame starting point</dfn> of a {{Node}} |N| in a
...
 Unstable Nodes {#sec-unstable-nodes}
 ------------------------------------
 
-A {{Node}} |N| <dfn export>has shifted</dfn> in a coordinate space |C| if the
-<a>starting point</a> of |N| in |C| differs from the <a>previous frame starting
-point</a> of |N| in |C| by 3 or more <a>pixel units</a> in either the horizontal
-or vertical direction.
+A {{Node}} |N| <dfn export>has shifted</dfn> in a coordinate space |C| if:
+
+* the <a>starting point</a> of |N| in |C| <a>differs significantly</a> from the
```
Comment (wangxianzhu) [inline index.bs:280]: I think we should consider transform-indifferent starting point for "move vector", too. The chromium CL does that, and it's not easy for it to consider transform-indifferent starting point when judging shift, but not when calculating move vector.

File: index.bs (hunk 1 above)
Comment (skobes) [reply index.bs:230]: I think "scroll indifference" is already captured by the consideration of ancestor scrolling-contents coordinate spaces in the definition of "unstable".

Comment (wangxianzhu) [reply index.bs:230]: Ack.

**Comment by skobes** [comment]
> The current CL uses the following algorithm for move vector:
> move_vector = starting_point - old_starting_point;
> transform_indifferent_move_vector = transform_indifferent_move_vector - old_transform_indifferent_move_vector;
> if (abs(transform_indifferent_move_vector.x()) < abs(move_vector.x()))
> move_vector.x() = transform_indifferent_move_vector.x();
> if (abs(transform_indifferent_move_vector.y()) < abs(move_vector.y()))
> move_vector.y() = transform_indifferent_move_vector.y();

Can't we just use move_vector then (if your CL knows both)?

I think the simplest approach is for the move vector to use only the "real" (i.e. transform-including) starting points, so that it corresponds most closely to what the user sees.

**Review by npm1 — APPROVED** [review]
LGTM modulo getting on the same page regarding the fix :)

**Comment by wangxianzhu** [comment]
> > The current CL uses the following algorithm for move vector:
> > move_vector = starting_point - old_starting_point;
> > transform_indifferent_move_vector = transform_indifferent_move_vector - old_transform_indifferent_move_vector;
> > if (abs(transform_indifferent_move_vector.x()) < abs(move_vector.x()))
> > move_vector.x() = transform_indifferent_move_vector.x();
> > if (abs(transform_indifferent_move_vector.y()) < abs(move_vector.y()))
> > move_vector.y() = transform_indifferent_move_vector.y();
> 
> Can't we just use move_vector then (if your CL knows both)?
> 
> I think the simplest approach is for the move vector to use only the "real" (i.e. transform-including) starting points, so that it corresponds most closely to what the user sees.

Using real starting points for move_vector SGTM.

The CL uses the algorithm so that PaintInvalidator can just pass one set of starting points to LayoutShiftTracker. LayoutShiftTracker then uses the starting points for both instability and shift distance. To calculate instability based on transform-indifferent starting points and shift distance based on real starting points, we need to add parameters in LayoutShiftTracker API. Will update the CL for this.

**Review by wangxianzhu — APPROVED** [review]
LGTM

**Pull request merged**